sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/healthcare/HL7v2IO.java

```diff
@@ -0,0 +1,658 @@
+        private transient AdaptiveThrottler throttler;
+
+        /** Instantiates a new Hl 7 v 2 message get fn. */
+        HL7v2MessageGetFn() {}
+
+        /**
+         * Instantiate healthcare client.
+         *
+         * @throws IOException the io exception
+         */
+        @Setup
+        public void instantiateHealthcareClient() throws IOException {
+          this.client = new HttpHealthcareApiClient();
+        }
+
+        /** Start bundle. */
+        @StartBundle
+        public void startBundle() {
+          if (throttler == null) {
+            throttler = new AdaptiveThrottler(1200000, 10000, 1.25);
+          }
+        }
+
+        /**
+         * Process element.
+         *
+         * @param context the context
+         */
+        @ProcessElement
+        public void processElement(ProcessContext context) {
+          String msgId = context.element();
+          try {
+            context.output(HL7v2Message.fromModel(fetchMessage(this.client, msgId)));
+          } catch (Exception e) {
+            failedMessageGets.inc();
+            LOG.warn(
+                String.format(
+                    "Error fetching HL7v2 message with ID %s writing to Dead Letter "
+                        + "Queue. Cause: %s Stack Trace: %s",
+                    msgId, e.getMessage(), Throwables.getStackTraceAsString(e)));
+            context.output(HL7v2IO.Read.DEAD_LETTER, HealthcareIOError.of(msgId, e));
+          }
+        }
+
+        private Message fetchMessage(HealthcareApiClient client, String msgId)
+            throws IOException, ParseException, IllegalArgumentException, InterruptedException {
+          final int throttleWaitSeconds = 5;
+          long startTime = System.currentTimeMillis();
+          Sleeper sleeper = Sleeper.DEFAULT;
+          if (throttler.throttleRequest(startTime)) {
+            LOG.info(String.format("Delaying request for %s due to previous failures.", msgId));
+            this.throttledSeconds.inc(throttleWaitSeconds);
+            sleeper.sleep(throttleWaitSeconds * 1000);
+          }
+
+          com.google.api.services.healthcare.v1alpha2.model.Message msg =
+              client.getHL7v2Message(msgId);
+
+          this.throttler.successfulRequest(startTime);
+          if (msg == null) {
+            throw new IOException(String.format("GET request for %s returned null", msgId));
+          }
+          this.successfulHL7v2MessageGets.inc();
+          return msg;
+        }
+      }
+    }
+  }
+
+  /** The type List HL7v2 message IDs. */
+  public static class ListHL7v2MessageIDs extends PTransform<PBegin, PCollection<String>> {
+
+    private final List<String> hl7v2Stores;
+    private final String filter;
+
+    /**
+     * Instantiates a new List HL7v2 message IDs with filter.
+     *
+     * @param hl7v2Stores the HL7v2 stores
+     * @param filter the filter
+     */
+    ListHL7v2MessageIDs(List<String> hl7v2Stores, String filter) {
```

Comment:
It might be worth supporting a the extra options on the new list API (view, and orderBy).  In particular, the view can return the full message payload right in the list, thereby saving a future retrieval step.

Reply:
Thanks for bringing this up.
Agreed on supporting getting the messages from the new list API to save the extra retrieval.

IMO this DoFn should remain as is to support backwards compatibility with the beta list api (which only has message IDs in the response) through a similar code path to the real time use cases which would subscribe to pubsub notifications (which also only have the message IDs and require retrieval).

@pabloem @lastomato thoughts on how we should best handle the two different APIs minimizing calls where possible?

I see two possible strategies:
1. Rely on user to use a different PTransform depending on which API version they are trying to hit.
   - For old api requiring "get" after list: `HL7v2IO.ListHL7v2MessageIDs` -> `HL7v2IO.Read`(current path).
   - Add a new PTransform for new api (where message contents are in the list): `HL7v2IO.V2.Read extends PTransform<PBegin, HL7v2Message>`
1. Restructuring so we can support both API responses with a single composite PTransform that follows this logic:
   - `ListMessageFn` makes the message list call and outputs to two `PCollection`s with the following logic:
        - IF response contains the `hl7V2Messages` key and either output the messages to the `PCollection<HL7v2Message>` with a `V2` tuple tag
        - ELSE  output the messageIDs `PCollection<String>` with a `V1` tuple tag
   - On the `V1` tagged PCollection apply `HL7v2IO.Read` to produce a `PCollection<HL7v2Message>`
   - `Flatten` the two PCollection<HL7v2Message> (one of which will always in theory be empty) to form your ultimate output collection

My thoughts of trade-offs are:
1. Pros: Simpler execution path, simpler to implement (strictly additive) given work already done. Cons: User has to know what they're doing and need to document / maintain both paths
1. Pros: Simpler interface for users Cons: more complicated execution path, heavier lift.

I'm torn on which would be preferable and would love input on which approach others think is cleanest before implementing. (Or suggestions to something I haven't thought of!)

Reply:
As a user -- I'll know what version of the V2 store I am using and likely am never switching back and forth.  For instance, I would just ignore the beta PTransform and use the alpha one.  Seems like that is not a large user burden.

Reply:
This can be resolved because beta API also supports view now.
Refactored to serve this use case w/ a separate PTransform and updated docstring.